lib/chef/provider/user/mac.rb

```diff
@@ -0,0 +1,607 @@
+        def create_user
+          cmd = [-"-addUser", new_resource.username]
+          cmd += ["-fullName", new_resource.comment]  if new_resource.property_is_set?(:comment)
+          cmd += ["-UID", new_resource.uid]           if new_resource.property_is_set?(:uid)
+          cmd += ["-shell", new_resource.shell]       if new_resource.property_is_set?(:shell)
+          cmd += ["-home", new_resource.home]         if new_resource.property_is_set?(:home)
+          cmd += ["-admin"]                           if new_resource.admin
+
+          # We can technically create a new user without the admin credentials
+          # but without them the user cannot enable SecureToken, thus they cannot
+          # create other secure users or enable FileVault full disk encryption.
+          if new_resource.property_is_set?(:admin_username) && new_resource.property_is_set?(:admin_password)
+            cmd += ["-adminUser", new_resource.admin_username]
+            cmd += ["-adminPassword", new_resource.admin_password]
+          end
+
+          # sysadminctl doesn't exit with a non-zero exit code if it encounters
+          # a problem. We'll check stderr and make sure we see that it finished
+          # correctly.
+          res = run_sysadminctl(cmd)
+          unless res.downcase =~ /creating user/
+            raise Chef::Exceptions::User, "error when creating user: #{res}"
+          end
+
+          # Wait for the user to be created and for dscl to flush everything
+          sleep 3
+
+          # Reload with up-to-date user information
+          @user_plist = load_user
+          reload_with_user_plist
+
+          converge_by("set password") { set_password } if new_resource.property_is_set?(:password)
+
+          if new_resource.manage_home
+            # "sydadminctl -addUser" will create the home directory if it's
+            # the default /Users/<username>, otherwise it sets it in plist
+            # but does not create it. Here we'll ensure that it gets created
+            # if we've been given a directory that is not the default.
+            unless ::File.directory?(new_resource.home) && ::File.exist?(new_resource.home)
+              converge_by("create home directory") do
+                shell_out!("createhomedir -c -u #{new_resource.username}")
+              end
+            end
+          end
+
+          if new_resource.property_is_set?(:gid)
+            # NOTE: Here we're managing the primary group of the user which is
+            # a departure from previous behavior. We could just set the
+            # PrimaryGroupID for the user and move on if we decide that actual
+            # group magement should be done outside of the core resource.
+            group_name, group_id, group_action = user_group_info
+
+            declare_resource(:group, group_name) do
+              members new_resource.username
+              gid group_id if group_id
+              action group_action
+              append true
+            end
```

Comment:
you probably want to `action :nothing` this one and then `.run_action(:create)` or whatever to force it to compile time.  you have a lot of converge_by blocks in here and this resource will run only after all of those complete.

Reply:
yeah `run_dscl` uses a `shell_out` and not an execute resource, so you're going to want to pick your poison and either do everything with chef sub-resources, including ruby_blocks where necessary, or else you're going to want to take all the sub-resources and force them to compile time.

this would be a good example of where #8668 would be useful.

Reply:
Ironically, the original pass at this used all sub-resources and then I changed it to just to the mixture :man_facepalming:

Reply:
yeah you gotta pick one or the other, the mixture crosses the streams.  if you can do it without any converge_by and use sub-resources entirely that's probably the best approach.  but then you still need to think about compile/converge mode.  when people start using converge_by because they want control over the output formatting and think that converge_by is equivalent to a "log" statement is when things usually get real weird...